### PR TITLE
Scheduling class support

### PR DESCRIPTION
**Pull request by js2xxx**
This PR adds support for scheduling classes.

Tracking issue: #1099.

### CODE REVIEW

**Comment by jellllly420** [comment]
Is it possible to break this PR into separate ones - one for the multi-scheduler mechanism and each of the others introducing a specific policy (on demand)? Currently I can't see through this PR 1) what is the right way to add our own scheduling policy and 2) why we need the real-time things now.

**Comment by js2xxx** [comment]
> 1) what is the right way to add our own scheduling policy and 2) why we need the real-time things now.

If my memory is correct, we don't need any mechanism for further customization of scheduling classes according to aforementioned plans and expectations, so simple & straightforward adaptations of the policies in the Linux kernel should be enough.

> Furthermore, with small and atomic PR's, it may be easier to debug and review.

It makes sense to some extent. However, the selection of scheduling policies by tasks depends on a single, unified `Priority`, which has already been used in other parts of the code base. That will complicate the handling code in the process of submitting individual PRs.

Moreover, it remains unresolve whether to continue to use `Priority` instead of a more specific `enum SchedPolicy` for scheduling policy selection, as stated in the RFC issue.

**Review by tatetian — COMMENTED** [review]
Good job! This is quite an important feature to Asterinas.

With that said, I am afraid that this PR is going to need some major revisions before getting merged. Here is my major concerns:

1. Need to be rebase to the latest main;
2. Add Rust docs to explain the scheduling strategies. You should not assume that every Asterinas developer would read Linux's schedulers before reading Asterinas's scheduler.
3. There are quite frequent locking and downcasting. I am wondering if this PR will be detrimental to the lmbench results.

File: kernel/src/sched/sched_class.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class.rs:136]: The atomic check & set on `task.schedule_info().cpu` should prevent this race. When a task is being enqueued into the scheduler, the `cpu` field will be set to its desired running cpu core after a successful check on whether its already in the scheduler. If its running cpu is not none (i.e. in the scheduler), the further procedures will not be executed.

File: kernel/src/sched/sched_class.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class.rs:94]: This is because methods on trait objects cannot be called by objects of its sub traits if `#[feature(trait_upcasting)]` is not enabled, which I didn't notice in `kernel/lib.rs` and wrote it by habit.

File: kernel/src/sched/sched_class.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class.rs:84]: Same as the reason of `as_any_*` in the definition of `SchedEntity`.

File: kernel/src/sched/sched_class.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class.rs:89]: Hand-written enum dispatching results in boilerplate codes, which I have already tried. If it is really critical regarding performance, I would recommend adding [`enum_dispatch`](https://docs.rs/enum_dispatch) to the crate dependencies so that the boilerplate could be avoided, or the manual implementation could be eventually taken into account.

File: kernel/src/sched/sched_class/fair.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs:231]: The current thread is popped out before task switching as stated in `pick_next`, so the update here doesn't affect the integrity of `BTreeSet`. A debug assertion could be added to ensure the invariant, but it would dramatically affect the debug performance and slow down the CI.

File: kernel/src/sched/sched_class/fair.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs:142]: The toolchain is too old, in which the methods like `lower_bound_mut` in `#![feature(btree_cursors)]` were not implemented by `BTreeSet` (which were merged [at the end of July, 2024](https://github.com/rust-lang/rust/pull/128309)). If I need to use `BTreeSet`, maybe thee toolchain should be updated first.

File: kernel/src/sched/sched_class/fair.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs:144]: I'll get rid of it since all the fields are to be mutated at the same time.

File: kernel/src/sched/sched_class/fair.rs
Comment (tatetian) [reply kernel/src/sched/sched_class/fair.rs]: Ok. In this case, write a TODO here and explain why `BTreeMap` is used when `BTreeSet` is more appropriate.

File: kernel/src/sched/sched_class.rs
Comment (tatetian) [reply kernel/src/sched/sched_class.rs]: So they can and should be removed, right?

File: kernel/src/sched/sched_class.rs
Comment (tatetian) [reply kernel/src/sched/sched_class.rs]: > Hand-written enum dispatching results in boilerplate codes

Could you elaborate on this point?  After removing `as_any` and other methods, `SchedEntity` provides almost no values. The `kind` method is provided inherently by using enums. So I can't see why dynamic dispatching on enums is needed.

File: kernel/src/sched/sched_class.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class.rs]: Yes. I've already removed it and `debug` in the latest commit.

File: kernel/src/sched/sched_class.rs
Comment (tatetian) [reply kernel/src/sched/sched_class.rs]: I will leave this issue open for now and reason the correctness of the scheduling logic after necessary Rust docs are added and the code is polished for readability.

File: kernel/src/sched/sched_class/fair.rs
Comment (tatetian) [reply kernel/src/sched/sched_class/fair.rs:231]: Ok. I will leave this issue open for now and reason about the correctness of the scheduling logic after necessary Rust docs are added and the code is polished for readability.

**Comment by tatetian** [comment]
As I saw a number of poorly-chosen names in this PR, I think it would benefit this PR to give two general pieces of advices on choosing good names.
1. Start function or method names with verbs. (Except for _trivial_ getter methods.)
2. Prefer accurate and specific names over inaccurate and vague names, even when this means the names become longer.

Here is a partial list of short-yet-vague names.

| Variable/method names | Locations             | Comments                                                                                                  |
|-----------------------|-----------------------|-----------------------------------------------------------------------------------------------------------|
| `entity`              | In `sched_class.rs`   | Rename it to `copy_sched_entity` for consistency with `SchedEntity` and emphasize the overhead of copying |
| `debug`               | In `SchedEntity`      | Rename it to `debug_info` to make it clear that debug is a noun nor verb                                  |
| `Wrapper`             | In `fair.rs`          | The name is not self-describing                                                                           |
| `r`                   | In `new_sched_entity` | I have no clue why a priority value is named `r`. Use `p` or `prio` instead                               |
| `consts`              | In `time.rs`          | The name and signature of `fn consts() -> (u64, u64)` is not helpful to readers at all                    |

Please fix these names as well as some other poor names that are not mentioned in the revision process. Thanks.

File: kernel/src/sched/sched_class.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class.rs]: `SchedEntity` it self is useless by now indeed. However, I think both `SchedEntity` and `SchedClassRq` should have the same dispatching approach to improve code consistency. So if We make `SchedEntity` to use enum dispatching, I would recommend making `SchedClassRq` to use it as well, which would result in the boilerplate code stated above.

Comment (js2xxx) [reply kernel/src/sched/sched_class.rs]: Hold on. I think I've found a better way to handle this.

File: kernel/src/sched/sched_class.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class.rs]: Renamed to a better name.

File: kernel/src/sched/sched_class.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class.rs]: Renamed.

File: kernel/src/sched/sched_class/mod.rs
Comment (tatetian) [inline kernel/src/sched/sched_class/mod.rs:69]: I found the interface of `SchedClassRq` trait weird. First, all current concrete runqueue types have an empty implementation for this `dequeue` method. So it is questionable whether having this method in the trait is necessary. Second, technically, the `pick_next` method already _dequeues_ the next thread. So the name of the `pick_next` method is misleading. Third, the `update_current` method is also weird. Because the "current" thread is already dequeued.

With that said, I don't have a good alternative design at hand. I just wanted to point out the problem. Maybe you would feel the same.

File: kernel/src/sched/sched_class/mod.rs
Comment (tatetian) [inline kernel/src/sched/sched_class/mod.rs:185]: I don't think the current implementation of `update_current` is correct since it does not respect the scheduling priority. Let's say the current task belongs to the fair scheduler. What if a timer interrupt occurs and a real-time thread is runnable? The real-time runqueue is not even looked.

**Review by tatetian — COMMENTED** [review]
Great improvements over the last version, especially the new comments added to explain the scheduling algorithms! I think the overall structure of the code is good. There is no doubt about that. But the implementation has still some rooms for improvement.

By the way, please remove intermediate commits and if necessary, break the PR into atomic commits.

File: kernel/src/sched/sched_class/mod.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/mod.rs:116]: Because we need to obtain the current CPU number in order to get the corresponding run queue and lock it.

File: kernel/src/sched/sched_class/mod.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/mod.rs:164]: The major difference IMO is that `_thread` only picks the next thread to run while `_current` furthers to set the current thread to it.

Comment (tatetian) [reply kernel/src/sched/sched_class/mod.rs:164]: The method names should be `pick_next_and_replace_current` and `pick_next`.

File: kernel/src/sched/sched_class/fair.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs:241]: `target_cpu` was originally used for migration between run queues. In this case for now, it IS wrong because no migration is available, so I'll wrap it in an option to keep the behavior expected.

File: kernel/src/sched/sched_class/fair.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs:32]: I have rewritten it with direct calculation and found the factor has always been 1.25 instead of 1.1 XoX, so I have changed the comment as well.

File: kernel/src/sched/sched_class/mod.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/mod.rs:164]: The current `pick_next_current` (gonna be replaced by `pick_next_and_replace_current`) is a method bound to `LocalRunQueue` trait, which is worth a second thought to be renamed.

Or, making both renamed methods independent and having the trait method simply forwarded to the actual implementation will do the thing as well.

**Comment by js2xxx** [comment]
The new `SchedAttr` results in little performance change:

    lat_ctx -P 1 2     4   8    16   32   64
    before:      0.62 1.40 1.53 1.73 1.86 1.88
    after:       0.62 1.43 1.59 1.66 1.79 1.86

File: kernel/src/sched/sched_class/fair.rs
Comment (tatetian) [inline kernel/src/sched/sched_class/fair.rs:106]: I am skeptical about the correctness of the current CFS implementation because of how `vruntime` and `period_start` values are updated and used.

First of all, after a task of `SCHED_OTHER` is spawned, the vruntime of this task is increased forever. Imagine this scenario: we have a task A that keeps running for a year. Then, a new task B is spawned. So the vruntime of task A will be much, much larger than that of task B. This means it would take a very loooong time before task A can be scheduled before task B. Even if task A is given the least nice value! Even if task A is I/O bound!

Second, the CFS algorithm is supposed to identify CPU-bound and I/O-bound tasks automatically by using vruntime and favor I/O-bound tasks over CPU-bound ones to reduce the latency of I/O. But the current CFS implementation does not behave this way. Consider the following situation: a CPU-bound task is the only running task on a CPU core. Then, an I/O-bound task wakes up and gets enqueued into a runqueue of a CFS scheduler. The I/O-bound task must have a small vruntime because it consumes lesser CPU time than the CPU-bound task. My understanding is that the expected behavior of CFS is to let the I/O-bound task (with a low vruntime) preempt the currently-running CPU-bound task (with a high vruntime) at the latest convenient moment (e.g., a scheduler tick).

```
    fn tick(
        &self,
        now: u64,
        start: u64,
        period_start: u64,
        total_weight: u64,
        period: u64,
    ) -> bool {
        // ...

        let slice = period * weight / total_weight;
        now - period_start > slice
}
```

But according to the current implementation (above), a preemption will only happen if the current running task consumes its assigned time slice. In other words, the I/O-bound task won't be able to preempt the CPU-bound one.

Third, I am confused with the semantics of the `period_start` attribute. Currently, this value of `period_start` is set when a task is _enqueued_ into a local runqueue.  (So this `period_start` should be called `enqueue_timestamp`.)

```rust
    fn enqueue_thread(&mut self, thread: &Arc<Thread>) {
        let attr = thread.sched_attr();

        attr.period_start.store(sched_clock(), Relaxed);
```

Then, the aforementioned tick method check if `now - period_start > slice`. Why do we want to calculate the elapsed time since the task is enqueued? This is NOT the consumed CPU time of this task. And then compare it with the assigned time slice? The code does not make sense to me at all.

File: kernel/src/sched/sched_class/mod.rs
Comment (tatetian) [inline kernel/src/sched/sched_class/mod.rs:230]: Avoid calling `sched_clock` twice.

File: kernel/src/sched/sched_class/mod.rs
Comment (tatetian) [inline kernel/src/sched/sched_class/mod.rs:291]: First of all, `update` is a poorly-chosen word. You can do a better job on naming.

Second, when `update` is true, we can simply short circuit the logic of "lookahead".

**Review by tatetian — COMMENTED** [review]
Here is my feedbacks for the latest reversion.

File: kernel/src/sched/sched_class/fair.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs:28]: It's because the actual factor is just 12.5% from the start, which I miscalculated it to 10%.

Comment (tatetian) [reply kernel/src/sched/sched_class/fair.rs:28]: Ok

File: kernel/src/sched/sched_class/fair.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs:74]: I cannot find anything which should be updated. Could you please point out which range of those texts are outdated?

File: kernel/src/sched/sched_class/fair.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs:106]: For possible starvation, yes, you're right. I've implemented a new mechanism `min_vruntime` to keep the values not too distant.

For responsiveness on I/O bound tasks, I've added another condition to trigger preemption:

    cur.vruntime >= rq.min_vruntime + normalized_slice

I've also refactored `tick` to ensure readability.

For `period_start`, yes, you're right. I've updated the period runtime counting algorithm.

Comment (tatetian) [reply kernel/src/sched/sched_class/fair.rs]: I have a working understanding of task scheduling, I am not an expert, especially when it comes to the intricacies of CFS. Implementing a robust and correct scheduling algorithm requires a level of detail that exceeds the time and resources I can currently dedicate. Therefore, I deeply value your expertise and professionalism in ensuring the accuracy and reliability of this implementation.

When I conduct code reviews, I do so under the assumption that the code is fundamentally correct, with the expectation that bugs are the exception rather than the norm. To this end, could I kindly ask you to perform an additional thorough review of the implementation? It has been three rounds of code review for this PR and I really hope we can merge it soon.

Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs]: I'm sorry for the trouble. I have been busy in the last two weeks, so the amount of time I've spent improving the quality of the PR is less than before. I hope it'll be merged soon from the bottom of my heart.

**Comment by tatetian** [comment]
@boterinas rerun

**Review by tatetian — COMMENTED** [review]
Thanks for revising the PR again. I think we are close to merge it. The code quality is now acceptable. I think the main obstacle is the correctness. Need to investigate why this PR causes more CI tests to fail than the current main. Also, the CFS algorithm may still have some flaws.

File: kernel/src/sched/sched_class/fair.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs:264]: The `min_queued_runtime` function is implemented as such to give some privilege to newly-initialized or long-waited threads. The subtractor is the normalized time slice of each thread in the run queue, whose meaning I think is described pretty clearly.

For the naming of `min_runtime`, the implementation of run queues per scheduling class doesn't include the current thread by design, so the names correspond to it.

**Comment by tatetian** [comment]
@lucassong-mh Could you investigate why the gvisor test failures in the latest CI? I found that some gvisor tests on the blocklist were run unexpectedly.
  * Failed alarm test: https://github.com/asterinas/asterinas/actions/runs/11855637762/job/33040263208?pr=1192#step:10:1107
  * Failed link tests: https://github.com/asterinas/asterinas/actions/runs/11855637762/job/33040263208?pr=1192#step:10:1795

**Comment by tatetian** [comment]
@cchanging Could you check out [the panic in `vm_mapping`](https://github.com/asterinas/asterinas/actions/runs/11855637762/job/33040262324?pr=1192#step:7:2292)? It seems to be an existing bug exposed by the new scheduler.

**Comment by tatetian** [comment]
@lucassong-mh Could you check out [the panic in Ext2 inode](https://github.com/asterinas/asterinas/actions/runs/11855637762/job/33040262905?pr=1192#step:9:2528)? It seems to be an existing bug exposed by the new scheduler.

File: kernel/src/sched/sched_class/fair.rs
Comment (tatetian) [reply kernel/src/sched/sched_class/fair.rs]: This is not how [Linux's CFS](https://github.com/torvalds/linux/blob/v2.6.24/kernel/sched_fair.c#L322-L325) maintains its `min_vruntime`.

```c
	/*
	 * maintain cfs_rq->min_vruntime to be a monotonic increasing
	 * value tracking the leftmost vruntime in the tree.
	 */
```

Our current `min_queued_vruntime` method does not return the _minimum_ vruntime of runnable tasks in a run queue, but the _second_ minimum vruntime _minus_ the nomalized time slice. This choice is really, really strange.

```rust
        // FIXME: Uncommenting this hangs the CI.
        // fair_attr
        //     .vruntime
        //     .fetch_max(self.min_queued_vruntime(), Relaxed);
```

We cannot take this value to update the vruntime of a newly-enqueued runnable task. Doing so may result the new task having the new minimum or the second minimum vruntime value in the run queue. This means that the `min_vruntime` of the runqueue may get decreased (the task's vruntime still increases due to `fetch_max`), breaking the monotonicity property.

In short, the `FairClassRq::min_vruntime` should keep the minimum vruntime of all runnable task, including the current one, and the value should be updated to ensure monotonicity.

In addition, the `vruntime` of a newly-enqueued task should be update to `rq.min_vruntime + penality` if `vurntime < rq.min_vruntime + penality`, where `penality` is a positive value controlled by the scheduler algorithm. See [the `place_entity` function](https://github.com/torvalds/linux/blob/49914084e797530d9baaf51df9eda77babc98fa8/kernel/sched_fair.c#L488).

**Comment by js2xxx** [comment]
@boterinas rerun

File: kernel/src/sched/sched_class/fair.rs
Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs]: If you stick to including the current running task in the calculation of `min_vruntime`, then we need to move the updating code into `update_current`, which again causes the time complexity to be `O(log(n))`. This is because we cannot maintain the leftmost reference to the run queue together with it (self-referencing) and can only retrieve it from the `BTreeSet` every time. Or we could use `BinaryHeap`, but efficiency is challenged.

Also, although maintaining the monotonicity of vruntime solves the hanging of the CI (the `test_linux` one), I cannot figure out its necessity. Anyway, let the CI do the talking.

Comment (tatetian) [reply kernel/src/sched/sched_class/fair.rs]: > Also, although maintaining the monotonicity of vruntime solves the hanging of the CI (the test_linux one), ...

Great. This is very encouraging news.

> ... I cannot figure out its necessity. Anyway, let the CI do the talking.

Let me put it this way and see if it makes sense to you. How do you define "fair" in CFS _mathematically_? Apparently, you can't define "fair" as always scheduling the task with the least vruntime. This is what the scheduler _does_, not what the tasks _obtain_.

In my (humble) opinion, CFS being "fair" means holding the following invariant 

$|T_i - T_j| <= P / N$

is held for any $1 \le i < j \le N$ at any point of time. Here, $T_i$ ($T_j$) denotes the vruntime of the $i$-th ($j$-th) task in a local run queue, $N$ the length of the run queue, and $P$ is the time of a scheduling period. A scheduling period is defined as a period of time during which each runnable task is scheduled _at least once_. I believe this invariant captures the intuition of being fair.

Put it another way, we have the following invariant:

$T_{max} - T_{min} <= P / N$

where $T_{min}$ and $T_{max}$ are the minimum and maximum vruntimes of tasks in the run queue. 

While the two invariants sound good in theory, one technical difficulty in practice is that tasks are not always runnable, making their vruntimes meaningless. For example, an I/O bound task may be blocked on an event for an indefinite amount of time. When it wakes up and joins the run queue again, its vruntime may be much smaller than the vruntimes of the ones in the queue. We cannot use its vruntime as is because it is not the scheduler's fault that the I/O bound task has a low runtime. So we must assign it a new value so that its vruntime can once again serve as an effective indicator of the CPU time consumed by the task. And when doing so, we must keep the aforementioned invariant. Assuming that the $T'$ is the updated vruntime of the newly-enqueued task, we must have

$T_{min} \le T' \le T_{max}$,

otherwise, we break the aforementioned two invariants once the task is added to the run queue.

Since $T' >= T_{min}$,  we know $T_{min}$ must be a _monotonic increasing value_.

Furthermore, since $T' >= T_{min}$, we can express it as

$T' = T_{min} + T_{penality}$

where $T_{penality} >= 0$ is a value chosen by the scheduler using some complicated heuristics.

The reasoning above is why I suggest maintaining the minimum vruntime (as Linux does) of the runqueue. 

> If you stick to including the current running task in the calculation of min_vruntime, then we need to move the updating code into update_current, which again causes the time complexity to be O(log(n))

What I insist are the invariants, not a particular implementation. Keeping the invariants does not dictate the $O(\log(N))$ complexity.

Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs]: > In my (humble) opinion, CFS being "fair" means holding the following invariant
> 
> | T i − T j | <= P / N

IMHO instead, fairness means every task gains indistingushable amount of vruntime in the same scheduling period. In other words:

$| \Delta T_{i} - P/N | < \epsilon$

Vruntimes are not necessarily equal, but their **differentials** should. The comparison of vruntimes themselves cannot determine the time slice they gain from each period. Hence, the absolute value of vruntime only determine the executing order in each period, and should consider not only weight of tasks, but also their requirements of responsiveness and other properties.

> What I insist are the invariants, not a particular implementation. Keeping the invariants does not dictate the O ( log ⁡ ( N ) ) complexity.

If you take the current task into account, then `update_current` should update `min_vruntime` as well:

```rust
let (vruntime, weight) = attr.fair.update(...);
self.min_vruntime = vruntime;
```

However, the code above is not correct if the current updated vruntime is greater then the minimal vruntime of ready-for-run tasks. So we should do this instead:

```rust
self.min_vruntime = self.threads.first()
    .map_or(vruntime, |item| vruntime.min(item.key()));
```

That is O(log N) in implementation if no additional structures are involved. There are alternatives though:

1. Use `BinaryHeap` instead, as accessing the top don't introduce `O(log n)` complexity, however the overall efficiency is challenged compared to B-trees.

2. Cloning the first item of `BTreeMap` and maintains it in the struct. This approach increases the memory overhead and adds logic maintained manually by developers, if it is considerable.

Note that keeping the current task in the B-tree or other ordered collections is ill-formed and causes unexpected behavior since its key (i.e. vruntime) should not be modified when the item is in the collection, which is an invariant that they depend on.

Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs]: Since the argument about theories cannot solve the current problems in the CI, I suggest we both postpone it until the issues are solved.

Comment (tatetian) [reply kernel/src/sched/sched_class/fair.rs]: Consider using [`RBTree`](https://docs.rs/intrusive-collections/latest/intrusive_collections/rbtree/struct.RBTree.html)?

Comment (js2xxx) [reply kernel/src/sched/sched_class/fair.rs]: > its key (i.e. vruntime) should not be modified when the item is in the collection, which is an invariant they depend on.

No, we can't. We can EITHER store `current` in the collection OR update its vruntime out of the collection — no other choice. It's simply because items in ordered collections are **ORDERED**. This variant cannot be violated.

**Comment by js2xxx** [comment]
The latest commit replaces B-trees with binary heaps. Let's see if the CI approves the change before benchmarking its performance.

**Comment by tatetian** [comment]
@js2xxx The bugfix for floating-point save/store is merged. It should resolve some of the CI failures that this PR encountered. Could you rebase this PR to the latest main branch?

**Comment by js2xxx** [comment]
It seems the CI passed on all tests for the first time :). @tatetian

**Comment by tatetian** [comment]
> It seems the CI passed on all tests for the first time :). @tatetian

Great news!

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Comment by tatetian** [comment]
@js2xxx Finally, this PR is ready to get merged. We all deserve a pat on the back. Need to fix the conflict. Thanks!

**Pull request merged**
